File: sverchok_nodes/renga/connection_protocol.py
# ##### BEGIN GPL LICENSE BLOCK #####
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# ##### END GPL LICENSE BLOCK #####

"""
Protocol for reliable TCP communication with Renga plugin
Uses length-prefixed messages (4 bytes big-endian + JSON data)
Compatible with C# Renga plugin protocol
"""

import struct
import socket
import json


def send_message(sock, message_dict):
    """
    Send a message with length prefix (4 bytes big-endian + JSON data)
    
    Args:
        sock: socket object
        message_dict: Message dictionary (will be converted to JSON)
    """
    if not sock:
        raise ValueError("Socket is None")
    
    # Convert to JSON
    json_str = json.dumps(message_dict)
    data = json_str.encode('utf-8')
    length = len(data)
    
    # Send length (4 bytes, big-endian)
    length_bytes = struct.pack('>I', length)
    sock.sendall(length_bytes)
    
    # Send data
    sock.sendall(data)


def receive_message(sock, timeout=10.0):
    """
    Receive a message with length prefix
    
    Args:
        sock: socket object
        timeout: timeout in seconds (default: 10.0)
    
    Returns:
        dict: Parsed JSON message as dictionary
    """
    if not sock:
        raise ValueError("Socket is None")
    
    # Set timeout
    sock.settimeout(timeout)
    
    # Read length (4 bytes)
    length_bytes = b''
    while len(length_bytes) < 4:
        chunk = sock.recv(4 - len(length_bytes))
        if not chunk:
            raise ConnectionError("Connection closed while reading message length")
        length_bytes += chunk
    
    length = struct.unpack('>I', length_bytes)[0]
    
    if length < 0 or length > 10 * 1024 * 1024:  # Max 10MB
        raise ValueError(f"Invalid message length: {length}")
    
    # Read JSON data
    buffer = b''
    while len(buffer) < length:
        chunk = sock.recv(length - len(buffer))
        if not chunk:
            raise ConnectionError("Connection closed while reading message data")
        buffer += chunk
    
    # Parse JSON and return as dict
    return json.loads(buffer.decode('utf-8'))

File: SverchokRenga/Connection/ConnectionProtocol.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace GrasshopperRNG.Connection
{
    /// <summary>
    /// Protocol for reliable TCP communication with length-prefixed messages
    /// </summary>
    public static class ConnectionProtocol
    {
        /// <summary>
        /// Send a message with length prefix (4 bytes big-endian + JSON data)
        /// </summary>
        public static async Task SendMessageAsync(NetworkStream stream, string json)
        {
            if (stream == null || !stream.CanWrite)
                throw new InvalidOperationException("Stream is not writable");

            var data = Encoding.UTF8.GetBytes(json);
            var length = BitConverter.GetBytes(IPAddress.HostToNetworkOrder(data.Length));
            
            // Send length (4 bytes)
            await stream.WriteAsync(length, 0, 4);
            
            // Send data
            await stream.WriteAsync(data, 0, data.Length);
            
            // Flush to ensure data is sent
            await stream.FlushAsync();
        }

        /// <summary>
        /// Receive a message with length prefix
        /// </summary>
        public static async Task<string> ReceiveMessageAsync(NetworkStream stream, int timeoutMs = 10000)
        {
            if (stream == null || !stream.CanRead)
                throw new InvalidOperationException("Stream is not readable");

            // Set read timeout
            stream.ReadTimeout = timeoutMs;

            // Read length (4 bytes)
            var lengthBytes = new byte[4];
            int totalRead = 0;
            
            while (totalRead < 4)
            {
                var read = await stream.ReadAsync(lengthBytes, totalRead, 4 - totalRead);
                if (read == 0)
                    throw new IOException("Connection closed while reading message length");
                totalRead += read;
            }
            
            int length = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(lengthBytes, 0));
            
            if (length < 0 || length > 10 * 1024 * 1024) // Max 10MB
                throw new IOException($"Invalid message length: {length}");

            // Read JSON data
            var buffer = new byte[length];
            totalRead = 0;
            
            while (totalRead < length)
            {
                var read = await stream.ReadAsync(buffer, totalRead, length - totalRead);
                if (read == 0)
                    throw new IOException("Connection closed while reading message data");
                totalRead += read;
            }
            
            return Encoding.UTF8.GetString(buffer, 0, length);
        }

        /// <summary>
        /// Synchronous version for compatibility
        /// </summary>
        public static void SendMessage(NetworkStream stream, string json)
        {
            if (stream == null || !stream.CanWrite)
                throw new InvalidOperationException("Stream is not writable");

            var data = Encoding.UTF8.GetBytes(json);
            var length = BitConverter.GetBytes(IPAddress.HostToNetworkOrder(data.Length));
            
            // Send length (4 bytes)
            stream.Write(length, 0, 4);
            
            // Send data
            stream.Write(data, 0, data.Length);
            
            // Flush to ensure data is sent
            stream.Flush();
        }

        /// <summary>
        /// Synchronous version for compatibility
        /// </summary>
        public static string ReceiveMessage(NetworkStream stream, int timeoutMs = 10000)
        {
            if (stream == null || !stream.CanRead)
                throw new InvalidOperationException("Stream is not readable");

            // Set read timeout
            stream.ReadTimeout = timeoutMs;

            // Read length (4 bytes)
            var lengthBytes = new byte[4];
            int totalRead = 0;
            
            while (totalRead < 4)
            {
                var read = stream.Read(lengthBytes, totalRead, 4 - totalRead);
                if (read == 0)
                    throw new IOException("Connection closed while reading message length");
                totalRead += read;
            }
            
            int length = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(lengthBytes, 0));
            
            if (length < 0 || length > 10 * 1024 * 1024) // Max 10MB
                throw new IOException($"Invalid message length: {length}");

            // Read JSON data
            var buffer = new byte[length];
            totalRead = 0;
            
            while (totalRead < length)
            {
                var read = stream.Read(buffer, totalRead, length - totalRead);
                if (read == 0)
                    throw new IOException("Connection closed while reading message data");
                totalRead += read;
            }
            
            return Encoding.UTF8.GetString(buffer, 0, length);
        }
    }
}



